cpp/include/raft/sparse/distance/detail/utils.cuh

```diff
@@ -37,6 +41,131 @@ inline int max_cols_per_block()
          sizeof(value_t);
 }
 
+template <typename value_idx, typename value_t>
+RAFT_KERNEL faster_dot_on_csr_kernel(value_t* __restrict__ dot,
+                                     const value_idx* __restrict__ indptr,
+                                     const value_idx* __restrict__ cols,
+                                     const value_t* __restrict__ A,
+                                     const value_t* __restrict__ B,
+                                     const value_idx nnz,
+                                     const value_idx n_rows,
+                                     const value_idx dim)
+{
+  auto vec_id  = threadIdx.x;
+  auto lane_id = threadIdx.x & 0x1f;
+
+  extern __shared__ char smem[];
+  value_t* s_A      = (value_t*)smem;
+  value_idx cur_row = -1;
+
+#pragma unroll
```

Comment:
Have you verified in the underlying SASS that this pragma unroll is doing anything? This does a static analysis when it's compiled and relies on compile-time constants to pull the runtime loops out into independent compiled loops. When the number of iterations is a runtime value, this will be silently skipped.

Reply:
Follow-up question- why are we iterating through all of the rows in each block? Shouldn't each block or warp be launched on an independent row? Otherwise, every block is going to duplicate the work, isn't it? 

Another thought- converting the CSR to a COO would be a more efficient way to do this, especially for CSR matrices of high sparsity- that way you could just have the kernel go through the edge list element-wise and perform the needed dots.

Reply:
Got it! I will remove some unuseful unroll, and I did try a COO version, but it is a little low efficient because of this memory-intensive kernel. The COO will increase the memory reading quantity

Reply:
>  and I did try a COO version, but it is a little low efficient because of this memory-intensive kernel. The COO will increase the memory reading quantity

That's not true, COO is for load balancing. Let's revisit this after 24.08.